src/profile-logic/js-tracer.js

```diff
@@ -441,3 +451,589 @@ export function getJsTracerLeafTiming(
+      if (generatedFuncIndex === undefined) {
+        // Create a new function only if the event string is different.
+        funcIndex = funcTable.length++;
+        funcTable.address.push(0);
+        funcTable.name.push(eventStringIndex);
+        funcTable.isJS.push(false);
+        funcTable.resource.push(-1);
+        funcTable.relevantForJS.push(true);
+        funcTable.fileName.push(null);
+        funcTable.lineNumber.push(null);
+        funcTable.columnNumber.push(null);
+
+        funcMap.set(eventStringIndex, funcIndex);
+      } else {
+        funcIndex = generatedFuncIndex;
+      }
+    }
+
+    // Try to find the prefix stack for this event.
+    let prefixIndex = unmatchedEventIndexes[unmatchedIndex];
+    while (prefixIndex !== null) {
+      const prefixEnd = unmatchedEventEnds[unmatchedIndex];
+      if (end <= prefixEnd) {
+        // This prefix is the correct one.
+        break;
+      }
+      // Keep on searching for the next prefix.
+      prefixIndex = unmatchedEventIndexes[unmatchedIndex];
+      unmatchedIndex--;
+    }
+
+    // Every event gets a unique frame entry.
+    const frameIndex = frameTable.length++;
+    frameTable.address.push(blankStringIndex);
+    frameTable.category.push(otherCategory);
+    frameTable.func.push(funcIndex);
+    // TODO - We could figure out the implementation, by tracking what the callee was.
+    frameTable.implementation.push(null);
+    frameTable.line.push(line);
+    frameTable.column.push(column);
+    frameTable.column.push(null);
+    frameTable.optimizations.push(null);
+
+    // Each event gets a stack table entry.
+    const stackIndex = stackTable.length++;
+    stackTable.frame.push(frameIndex);
+    stackTable.category.push(otherCategory);
+    stackTable.prefix.push(prefixIndex);
+    stackMap.set(tracerEventIndex, stackIndex);
+
+    // Add this stack to the unmatched stacks of events.
+    unmatchedIndex++;
+    if (unmatchedIndex === 0) {
+      // Reached a new root, reset the index to 1.
+      unmatchedIndex = 1;
+    }
+    unmatchedEventIndexes[unmatchedIndex] = tracerEventIndex;
+    unmatchedEventEnds[unmatchedIndex] = end;
+  }
+
+  return { thread, stackMap };
+}
+
+type JsTracerFixed = {|
+  events: Array<IndexIntoStringTable>,
+  start: Array<Microseconds>,
+  end: Array<Microseconds>,
+  line: Array<number | null>, // Line number.
+  column: Array<number | null>, // Column number.
+  length: number,
+|};
+
+/**
+ * JS Tracer information has a start and duration, but due to precision issues, this
+ * can lead to errors with computing parent/child relationships. This function converts
+ * it into a start and end time, and corrects issues where the events incorrectly
+ * overlap.
+ */
+export function getJsTracerFixed(jsTracer: JsTracerTable): JsTracerFixed {
+  if (jsTracer.length === 0) {
+    return {
+      events: [],
+      start: [],
+      end: [],
+      line: [],
+      column: [],
+      length: 0,
+    };
```

Comment:
Ah, I started doing this but the realized this is the transitory "fixed" variant, where it has a `start` and `end` rather than `start` and `duration`. I don't think it's worth adding a function since this is the one place it'll happen. I added a clarifying comment though.